apps/aecontract/src/aect_state_tree.erl

```diff
@@ -226,21 +210,18 @@ verify_poi(Pubkey, Contract, Poi) ->
     end.
 
 verify_store_poi(Id, Store, IsLegalStoreFun, Poi) ->
-    %% Check separately absence of (invalid) empty key because Merkle
-    %% tree iterator-from-key, used in PoI lookup, does not return
-    %% initial key if present.
-    case aec_poi:lookup(Id, Poi) of
-        {ok, _StoreValWithEmptyKey} -> {error, bad_proof};
-        {error, not_found} ->
-            case lookup_store_poi(Id, Poi) of
-                {ok, Store} ->
-                    case IsLegalStoreFun(Store) of
-                        true -> ok;
-                        false -> {error, bad_proof}
-                    end;
-                {ok, _} -> {error, bad_proof};
-                {error, _} = E -> E
-            end
+    StoreMap = aect_contracts_store:subtree(<<>>, Store),
+    try lookup_store_poi(Id, Poi) of
+        {ok, Store1} ->
+            StoreMap1 = aect_contracts_store:subtree(<<>>, Store1),
+            case StoreMap =:= StoreMap1 andalso IsLegalStoreFun(Store) of
```

Comment:
I think this is not good enough. The test is supposed to check that the store is "complete", i.e., that the MPT is complete for the store. Calling `aect_contracts_store:subtree(Store)` should be good enough. Or even better, make another function `aect_contracs_store:verify/1` that just traverses the store and throws away the result.

Unless there are some invariants on the values as well, in which case you could check them as well.

Reply:
I don't understand. Before, the store in the poi was a `binary() => binary()` map, and we checked that you got the same map back when looking up the poi. Now, due to the write cache, you might not have exactly the same representation so I convert to map and then check.

Or are you talking about the `IsLegalStoreFun` part?